### PR TITLE
Add support for TouchEvent

### PR DESCRIPTION
**Pull request by rickyhan**
#304

### CODE REVIEW

**Comment by koute** [comment]
Hmm... the failure on Travis is interesting; it might be related to the new Emscripten which the newest `cargo-web` is using. You can ignore it.

File: src/webapi/events/touch.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
+use webcore::value::Reference;
+use webcore::try_from::TryInto;
+use webapi::event::{IEvent, IUiEvent, UiEvent, Event};
+use webapi::touch::Touch;
+
+/// The `ITouchEvent` interface represents events that occur due to the user
+/// interacting with a touch device (such as a phone).
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent)
+// https://w3c.github.io/uievents/#touchevent
```
Comment (koute) [inline src/webapi/events/touch.rs]: These links are not valid. (:

In this case this link should be:

https://w3c.github.io/touch-events/#idl-def-touchevent

Please fix the other links too.

If it's not possible to link to something (e.g. I think the fields in the `TouchEvent` interface cannot be linked to?) then you can skip those.

File: src/webapi/touch.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+impl Touch {
+
+    /// Returns a unique identifier for this Touch object. A given touch point
+    /// (say, by a finger) will have the same identifier for the duration of
+    /// its movement around the surface. This lets you ensure that you're
+    /// tracking the same touch all the time.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Touch/identifier)
+    #[inline]
+    pub fn identifier(&self) -> f64 {
```
Comment (koute) [inline src/webapi/touch.rs]: According to the specs this is `long` (so `i32`), not `f64`.

File: src/webapi/touch.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        js!(
+            return @{self.as_ref()}.pageY;
+        ).try_into().unwrap()
+    }
+
+    /// Returns the Element on which the touch point started when it was first placed on the surface, even if the touch point has since moved outside the interactive area of that element or even been removed from the document.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Touch/target)
+    #[inline]
+    pub fn target(&self) -> Element {
```
Comment (koute) [inline src/webapi/touch.rs]: According to specs this returns `EventTarget`, not `Element`.

(If you really want to add a convenance method which returns an `Element` you can add `fn target_element( &self ) -> Option< Element >` if you want.)

File: src/webapi/touch.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        js!(
+            return @{self.as_ref()}.target;
+        ).try_into().unwrap()
+    }
+
+    /// Returns the X radius of the ellipse that most closely circumscribes the area of contact with the screen. The value is in pixels of the same scale as screenX.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Touch/radiusX)
+    #[inline]
+    pub fn radius_x(&self) -> f64 {
```
Comment (koute) [inline src/webapi/touch.rs]: Please leave a `TODO` here that this (and some of the others) should probably be changed to `f32` at a later date.

(Currently it doesn't matter as everything goes through the JS so it's ~turtles~ `f64` all the way down, but the specs say that those should be `f32` which I think will make a difference when we get native DOM bindings for WASM.)

**Comment by koute** [comment]
Did you close this PR by a mistake? (:

File: src/webapi/touch.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        js!(
+            return @{self.as_ref()}.target;
+        ).try_into().unwrap()
+    }
+
+
+
+    /// Returns the X radius of the ellipse that most closely circumscribes the area of contact with the screen. The value is in pixels of the same scale as screenX.
+    ///
+    /// TODO: f32 is not yet supported
```
Comment (koute) [inline src/webapi/touch.rs]: Please move this TODO out of the doc comment; it's a TODO for us, not necessarily for the users. (:

File: src/webapi/touch.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+impl Touch {
+
+    /// Returns a unique identifier for this Touch object. A given touch point
+    /// (say, by a finger) will have the same identifier for the duration of
+    /// its movement around the surface. This lets you ensure that you're
+    /// tracking the same touch all the time.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Touch/identifier)
+    #[inline]
+    pub fn identifier(&self) -> i64 {
```
Comment (koute) [inline src/webapi/touch.rs]: Why `i64`? (: The specs say that it's `long`, which is supposed to be `i32`, so is the `i64` here by mistake or is there an actual reason for it?

File: src/webapi/touch.rs [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    Reference,
+    Value,
+};
+
+/// The Touch interface represents a single contact point on a touch-sensitive
+/// device. The contact point is commonly a finger or stylus and the device may
+/// be a touchscreen or trackpad.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Touch)
+#[derive(Clone, Debug, Eq, PartialEq, ReferenceType)]
```
Comment (koute) [inline src/webapi/touch.rs]: Please put this specs link here:

https://w3c.github.io/touch-events/#touch-interface

You don't have to add links to the methods as it's not possible to link to the fields anyway.

**Comment by koute** [comment]
@rickyhan Can you confirm that you don't intend to continue these PRs? (It's perfectly fine if you don't. I just need to know so that I can reuse the work you've already done.)

**Comment by KappaDistributive** [comment]
What's stopping this PR from getting pulled? 

The only concerns left seem to be the return type of identifier() and a missing reference to https://w3c.github.io/touch-events/#touch-interface.

If there's more work to do, I'd be happy to finish it.

**Comment by koute** [comment]
@KappaDistributive Yes, you're right that it's almost finished. I was planning on pulling it in very soon myself; I just had some other stuff to finish. I guess I can do this sooner rather than layer. (:

**Comment by koute** [comment]
Pushed to `master`. I'll probably release it tomorrow.

**Comment by KappaDistributive** [comment]
Awesome. Thanks!

**Pull request closed without merging**